Write-up: Building Privacy-Protecting Infra (#97)
vacp2p/vac.dev: research/2022-11-02-building-privacy-protecting-infrastructure.md

privacy-protection guarantees would be very useful for high-value targets.

This doesn't begin to touch on the so called "dapps" that make a lot of
sacrifices in how they function, from the way domains work, to how websites are
hosted and the reliance on centralized services for communication. We see this
time and time again, where centralized, single points of failure systems works
for a while, but then eventually fails.

In many cases an individual user might not care enough though, and for platforms
the lure to take shortcuts is strong. That is why it is important to be
principled, but also pragmatic in terms of the trade-offs that you allow on top.
We'll touch more on this in the design goals around modularity that Waku has.

### ZK for Privacy-protecting infrastructure

ZKPs are a wonderful new tool. Just like smart contracts enables programmable
money, ZKPs allow us to express fundamentally new things. In line with the great
tradition of trust-minimization, we can prove statement while revealing the
absolute minimum information necessary. This fits the definition of privacy, the
power to selectively reveal yourself, perfectly. I'm sure I don't need to tell

> anyone in this room but this is truly revolutionary. The technology is advancing

Author: good chat

extremely fast and often it is our imagination that is the limit.

![Zero knowledge](/img/building_private_infra_zk.png)

### Waku

What is Waku? It is a set of modular protocols for p2p communication. It has a
focus on privacy, security and being able to run anywhere. It is the spiritual
success to Whisper.

By modular we mean that you can pick and choose protocols and how you use them
depending on constraints and trade-offs. For example, bandwidth usage vs
privacy.

It is designed to work in resource restricted environments, such as mobile
phones and in web browsers. It is important that infrastructure meets users
where they are and supports their real-world use cases. Just like you don't need
your own army and a castle to have your own private bathroom, you shouldn't need
to have a powerful always-on node to get reasonable privacy and
censorship-resistance. We might call this self-sovereignty.

> ### Waku - adaptive nodes

Reviewer: Just to capture the service network idea in the title:
```suggestion
### Waku - adaptive nodes/service network
```

One way of looking at Waku is as an open service network. There are nodes with
varying degrees of capabilities and requirements. For example when it comes to
bandwidth usage, storage, uptime, privacy requirements, latency requirements,
and connectivity restrictions.

We have a concept of adaptive nodes that can run a variety of protocols. A node
operator can choose which protocols they want to run. Naturally, there'll be
some nodes that do more consumption and other nodes that do more provisioning.
This gives rise to the idea of a service network, where services are provided
for and consumed.

![Adaptive Nodes](/img/building_private_infra_adaptive.png)

### Waku - protocol interactions 

> There are many protocols that interact. Waku Relay protocol is based libp2p

Reviewer 2: ```suggestion
There are many protocols that interact. Waku Relay protocol is based on libp2p
```

> GossipSub for p2p messaging. We have filter for bandwidth-restricted nodes to

Reviewer: I'd put links to the rfc of the protocols mentioned in here like filter, lightpush, etc.

only receive subset of messages. Lightpush for nodes with short connection

> windows to push messages into network. Store for nodes that want to retrieve

Reviewer: ```suggestion
windows to push messages into network. Store protocol for nodes that want to retrieve
```

historical messages.

On the payload layer, we provide support for Noise handshakes/key-exchanges.

> This means that as a developers, you can get end-to-end encryption and expected

Reviewer: ```suggestion
This means that as a developer, you can get end-to-end encryption and expected
```

guarantees out of the box. We have support for setting up a secure channel from

> scratch, and all of this paves the way for providing Signal's Double Ratchet at

Reviewer: A link to the Signal's Double Ratchet would be great.

the protocol level much easier. We also have experimental support for
multi-device usage. Similar features have existed in for example the Status app
for a while, but with this we make it easier for any platform using Waku to use
it.

There are other protocols too, related to peer discovery, topic usage, etc. See
[specs](https://rfc.vac.dev/) for more details.

![Protocol Interactions](/img/building_private_infra_interactions.png)

> ### Waku - Network

Reviewer: ```suggestion
### Waku - network challenges
```

these.

The spam problem arises on the gossip layer when anyone can overwhelm the
network with messages. The service incentivization is a problem when nodes don't
directly benefit from the provisioning of a certain service. This can happen if
they are not using the protocol directly themselves as part of normal operation,
or if they aren't socially inclined to provide a certain service. This depends a
lot on how an individual platform decides to use the network.

![Waku Network](/img/building_private_infra_network.png)

### Dealing with network spam and RLN Relay

Since the p2p relay network is open to anyone, there is a problem with spam. If
we look at existing solutions for dealing with spam in traditional messaging
systems, a lot of entities like Google, Facebook, Twitter, Telegram, Discord use
phone number verification. While this is largely sybil-resistant, it is
centralized and not private at all.

Historically, Whisper used PoW which isn't good for heterogenerous networks.
Peer scoring is open to sybil attacks and doesn't directly address spam
protection in an anonymous p2p network.

The key idea here is to use RLN for private economic spam protection using
zkSNARKs.

> I'm not going to go into too much detail of RLN here. If you are interested, I

Reviewer: How about mentioning the published papers as well as the rln and rln-relay specs in here.

Author: RLN spec ok

gave a [talk](https://www.youtube.com/watch?v=g41nHQ0mLoA) in Amsterdam at
Devconnect about this. We also have some write-ups on
[here](https://vac.dev/rln-relay) Sanaz who has been pushing a lot of this from
our side. There's also another talk at Devcon by Taylor going into RLN in more
detail.

I'll briefly go over what it is, the interface and circuit and then talk about
how it is used in Waku.

### RLN - Overview and Flow

RLN stands for Rate Limiting Nullifier. It is an anonyomous rate limiting
mechanism based on zkSNARKs. By rate limiting we mean you can only send N
messages in a given period. By anonymity we mean that you can't link message to
a publisher. We can think of it as a voting booth, where you are only allowed to
vote once every election.

![Voting Booth](/img/building_private_infra_vote.png)

It can be used for spam protection in p2p messaging systems, and also rate
limiting in general, such as for a decentralized captcha.

> There are three parts to it. You register somewhere, then you can signal and

Reviewer: ```suggestion
There are three parts to it. An RLN group to which you need to register, then you can signal by proving that you are a member of the RLN group and
```

finally there's a verification/slashing phase. You put some capital at risk,

> either economic or social, and if you double signal you get slashed.

Reviewer: ```suggestion
either economic or social, and if you double signal, your membership credentials get disclosed and you get slashed.
```

### RLN - Circuit

> Here's what the private and public input to the circuit looks like. The identity

Reviewer: ```suggestion
Here's what the private and public inputs to the circuit look like. The identity
```

secret is generated locally, and we create an identity commitment that is

> inserted into a merkle tree. We then use merkle proofs to prove membership.

Reviewer: ```suggestion
inserted into a Merkle tree. We then use Merkle proofs to prove membership.
```

Registered member can only signal once for a given epoch or external nullifier,
for example every ten seconds in Unix time. RLN identifer is for a specific RLN
app.

We also see what the circuit output looks like. This is calculated locally. `y`
is a share of the secret equation, and the (internal) nullifier acts as a unique
fingerprint for a given app/user/epoch combination. How do we calculate `y` and
the internal nullifier?

> ```

Reviewer: Not sure, but something like this may work (or any language close to circom in terms of syntax)
```suggestion
```circom
```

// Private input
signal input identity_secret;
signal input path_elements[n_levels][1];
signal input identity_path_index[n_levels];

// Public input
signal input x; // signal_hash
signal input epoch; // external_nullifier
signal input rln_identifier;

// Circuit output
signal output y;
signal output root;
signal output nullifier;
```

<!-- ![Circuit](/img/building_private_infra_circuit.png) -->

### RLN - Shamir's secret sharing

This is done using Shamir's secret sharing. Shamir’s secret sharing is based on
idea of splitting a secret into shares. This is how we enable slashing of funds.

> In this case we have two shares. If a given identity a0 signals twice in

Reviewer: ```suggestion
In this case, we have two shares. If a given identity `a0` signals twice in
```

> epoch/external nullifier, a1 is the same. For a given RLN app,

Reviewer: ```suggestion
epoch/external nullifier, `a1` is the same. For a given RLN app,
```

is worth noting that this can be combined with peer scoring, for example for
duplicate messages or invalid ZK proofs.

In line of Waku's goals of modularity, RLN Relay is applied on a specific subset
of pubsub and content topics. You can think of it as an extra secure channel.

![RLN Relay](/img/building_private_infra_rlnrelay.png)

### RLN Relay cross-client testnet

Where are we with RLN Relay deployment? We've recently launched our second
testnet. This is using RLN Relay with a smart contract on Goerli. It integrates
with our example p2p chat application, and it does so through three different
clients, nwaku, go-waku and js-waku for browsers. This is our first p2p
cross-client testnet for RLN Relay.

Here's a screenshot of a short demo. This shows a user registering in a browser,
signaling through JS-Waku. It then gets relayed to a nwaku node, that verifies
the proof. When more than one message is sent in a given epoch, it detects it as
spam and discards it. Slashing hasn't been implemented fully yet in the client
and is a work in progress.

If you are curious and want to participate, you can join the effort on our [Vac

> Discord(https://discord.gg/PQFdubGt6d). We also have

Reviewer 3: ```suggestion
Discord](https://discord.gg/PQFdubGt6d). We also have
```

Not sure the line breaks in the midfle of the link would work as well

verification code, circuits etc.  Rust strengths is that it is systems-based and
easy to interface with other language runtime such as Nim, Go, Rust, C. It also
gives access to other Rust ZK ecosystems such as arkworks. This opens door for
using other constructs, such as Halo2. This becomes especially relevant for
constructs where you don't want to do a trusted setup or where circuits are more
complex/custom and performance requirements are higher.

In general with Zerokit, we want to make it easy to build and use ZKP in a
multitude of environments, such as mobile phones and web browsers. Currently it
is too complex to write privacy-protecting infrastructure with ZKPs considering
all the languages and tools you have to learn, from JS, Solidity and Circom to
Rust, WASM and FFI. And that isn't even touching on things like secure key
storage or mobile dev. Luckily more and more projects are working on this,
including writing DSLs etc. It'd also be exciting if we can make a useful
toolstack for JS-less ZK dev to reduce cognitive overhead, similar to what we
have with something like Foundry.

### Other research

I also want to mention a few other things we are doing. One thing is

> [https://rfc.vac.dev/](protocol specifications). We think this is very important

Reviewer 3: ```suggestion
[protocol specifications](https://rfc.vac.dev/). We think this is very important
```

for p2p infra, and we see a lot of other projects that claim to do it p2p
infrastructure but they aren't clear about guarantees or how stable something
is. That makes it hard to have multiple implementations, to collaborate across
different projects, and to analyze things objectively.

Related to that is publishing [papers](https://vac.dev/publications). We've put
out three so far, related to Waku and RLN-Relay. This makes it easier to
interface with academia. There's a lot of good researchers out there and we want
to build a better bridge between academia and industry.

Another thing is [network](https://vac.dev/wakuv2-relay-anon)
[privacy](https://github.com/vacp2p/research/issues/107u). Waku is modular with
respect to privacy guarantees, and there are a lot of knobs to turn here
depending on specific deployments. For example, if you are running the full
relay protocol you currently have much stronger receiver anonymity than if you

> are running filter protocol from a bandwidth or connectivity restricted node.

Reviewer: ```suggestion
are running filter protocol from a bandwidth or connectivity-restricted node.
```

We aim to make this pluggable depending on user needs. E.g. mixnets such as Nym
come with some trade-offs but are a useful tool in the arsenal. A good mental
model to keep in mind is the anonymity trilemma, where you can only pick 2/3 out
of low latency, low bandwidth usage and strong anonymity.

We are currently exploring [Dandelion-like
additions](https://github.com/vacp2p/research/issues/119) to the relay/gossip
protocol, which would provide for stronger sender anonymity, especially in a
multi-node/botnet attacker model. As part of this we are looking into different
parameters choices and general possibilities for lower latency usage. This could

> make it more amendable for latency sensitive environments, such as validator

Reviewer 2: `amenable` rather than `amendable`? Or maybe you do mean it can be amended.
